### PR TITLE
Remove constraint on the width of the gutter function

### PR DESCRIPTION
**Pull request by irevoire**
Hello @archseer, I was looking at your PR, and I wondered if it would not be possible to remove the need for a fixed size width entirely?

By doing something like that, we could create other gutters more easily; what do you think?
For example, a `git blame` gutter would not need to preload all the names appearing in the file to pre-compute a maximum width for the current view.
Also, for the diagnostic gutter: if there is no diagnostic, the unused column gets automatically removed without any action.

### CODE REVIEW

**Comment by archseer** [comment]
We need to be able to determine the gutter width for calculations like `view.inner_area`.

> Also, for the diagnostic gutter: if there is no diagnostic, the unused column gets automatically removed without any action.

I find that a bit problematic in vim: without setting a fixed margin for the signs column, it will shift left and right as you type (causing code to do the same)

**Comment by pickfire** [comment]
> I find that a bit problematic in vim: without setting a fixed margin for the signs column, it will shift left and right as you type (causing code to do the same)

I find that annoying too. At the same time I also don't like the gutter wasting a single column causing my 80 characters terminal to not display a character on the right, I wish there is a way to disable the gutter entirely. But I do think if it's enabled a fixed gutter is good.

**Comment by sudormrfbin** [comment]
> I find that a bit problematic in vim: without setting a fixed margin for the signs column, it will shift left and right as you type (causing code to do the same)

I fixed that in vim by coloring the line numbers themselves with with diagnostic colors like red and yellow, though not everyone might like it.

**Comment by archseer** [comment]
Neovim had an interesting option to display the signs column instead of the line number, if a sign was present on that line. It would get a bit messy if you had a lot of signs, but it saved that extra column.

**Comment by irevoire** [comment]
> I find that a bit problematic in vim: without setting a fixed margin for the signs column, it will shift left and right as you type (causing code to do the same).

I think the good thing with functions is that we can define a generic `fixed_size` function that takes any gutter function and an `usize` / `u16` and fix its size.
And then the day we let people configure their gutter as they want, it would be so great to have some kind of syntax to let them do what they want easily, you could have that in your `config.toml` for example:
```
gutters =  "{diagnostic:1}{numbers:>5}|"
```
Or the conf I want:
```
gutters = "{diagnostic:1}{number}|"
```

**Comment by irevoire** [comment]
> We need to be able to determine the gutter width for calculations like view.inner_area.
I tried to propagate the `gutter_width` in the view and then everywhere needed as you said in [this](https://github.com/helix-editor/helix/pull/783#discussion_r717210923) comment.
Sadly I don't really know what the rendering pipeline is and encountered some mutability issues. Currently, I solved everything with a `Cell`, but that's really suboptimal :grimacing: 
You can look at this commit if you have the time :grin: : https://github.com/helix-editor/helix/pull/790/commits/410c1b5efe70db14ba7c291d4adc5b4084daca04

**Comment by irevoire** [comment]
And here is an implementation of my `fixed_width_gutter` function, and a `span` function to insert a space where I want :grin: 
https://github.com/helix-editor/helix/pull/790/commits/649861e4d09ce2ebb58d969616574df1e4438176

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -437,52 +423,145 @@ impl EditorView {
...
-                    "●",
-                    1,
-                    match diagnostic.severity {
-                        Some(Severity::Error) => error,
-                        Some(Severity::Warning) | None => warning,
-                        Some(Severity::Info) => info,
-                        Some(Severity::Hint) => hint,
-                    },
-                );
+        fn fixed_width_gutter(
```
Comment (AloeareV) [inline helix-term/src/ui/editor.rs:426]: I can't think of a situation where I've seen function-scoped functions. I find it makes it really hard to read what's going on in `render_gutter`, as the logic is interrupted by over 100 lines of helper-function definition.

**Comment by archseer** [comment]
Not sure why this got closed 😮

**Comment by cole-h** [comment]
Likely because it was targeting your `gutter_fn` branch, which you deleted after merging :P

**Comment by archseer** [comment]
It's set to auto delete. Usually github will re-target the PR to whatever branch `gutter-fn` merged into

**Pull request closed without merging**